### PR TITLE
Mapped Items not visible in Collection browse

### LINKED ISSUE #3141: Mapped Items not visible in Collection browse
**Describe the bug**
Discovered while testing https://github.com/DSpace/dspace-angular/pull/983.  While Items can be mapped successfully to different Collections, they **never** seem to appear in the Collection browse listing.  They are, however, visible in the Collection's "Item Mapper" tab (on the Edit Collection screen) when using https://github.com/DSpace/dspace-angular/pull/983.

**To Reproduce**
1. This is most easily tested by installing https://github.com/DSpace/dspace-angular/pull/983
2. Edit a Collection and Map a few items to that Collection.
3. Now, visit the Collection homepage (mapped items will not be listed).  Under that Collection, click on Browse "By Title", "By Issue Date", "By Author" or "By Subject"... the Mapped Item will also not appear in any of those browses.

My guess is that Mapped Items are not being properly added/updated in our Discovery/Solr indexes.

**Expected behavior**
Obviously, an Item mapped into a Collection should appear in that Collection's browse pages.

### PR DESCRIPTION
**Pull request by jonas-atmire**
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #3141 
* Related to [Angular feature](https://github.com/DSpace/dspace-angular/pull/983)

## Description
When mapping/unmapping items to a collection, the actual indexes of the items were not updated.
This had the side-effect, that these items couldn't be found in the browse/search of the collection that they were mapped to. (Because the search uses a scope, that can't find the item in solr, since it doesn't have the correct location information in the solr record)
Without any codechanges, this could already be fixed by pushing a metadata edit on the item, which indexes the item (Together with the mapped collection information). Making sure that the collection map / unmap 

## Instructions for Reviewers
Please add a more detailed description of the changes made by your PR. At a minimum, providing a bulleted list of changes in your PR is helpful to reviewers.

List of changes in this PR:
* As mentioned in the description, the actual indexing of the item fixes this problem.
* Since the code already created a new context event when [adding a new item](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/content/CollectionServiceImpl.java#L617) or removing a [previous one](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/content/CollectionServiceImpl.java#L638), the only change that needed to happen is to make sure that the underlying code knew that the item was actually changed
    * Which would turn trigger the discovery indexing consumer.
    * As such, we added a call to the 'setModified' method, to make sure the item was marked as 'updated' 

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

1. This is most easily tested by installing / setting up the latest Angular commit together with this PR
2. Edit a Collection and Map a few items to that Collection.
3. Now, visit the Collection homepage (mapped items will not be listed). Under that Collection, click on Browse "By Title", "By Issue Date", "By Author" or "By Subject"... the Mapped Item will **appear** correctly in those browses.
4. Remove the mapping for the previous items again
5. The items should **not** appear in the aforementioned browses

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/content/CollectionServiceImpl.java [hunk 1]
```diff
@@ -612,6 +612,7 @@ public void addItem(Context context, Collection collection, Item item) throws SQ
         // Instead we add the collection to an item which works in the same way.
         if (!item.getCollections().contains(collection)) {
             item.addCollection(collection);
+            item.setModified();
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/content/CollectionServiceImpl.java:615]: I would prefer to modify the IndexConsumer to deal properly with the ADD event in a collection instead than flag the item is updated when it is indeed potentially unchanged. This avoid the need to reprocess the item by other consumer that are not really interested in the collection mapping and could lead to more efficient implementation

File: dspace-api/src/main/java/org/dspace/content/CollectionServiceImpl.java [hunk 2]
```diff
@@ -632,7 +633,7 @@ public void removeItem(Context context, Collection collection, Item item)
         } else {
             //Remove the item from the collection if we have multiple collections
             item.removeCollection(collection);
-
+            item.setModified();
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/content/CollectionServiceImpl.java:636]: see previous, comment. The REMOVE event from the collection can be used

**Review by abollini — CHANGES_REQUESTED** [review]
thanks @jonas-atmire for this PR. The test looks good to me so they are sufficient to verify that the solution works as expected. I would prefer a different approach than emit a item modification since it should be sufficient to deal with the already existing collection events (see inline comments)

**Review by tdonohue — APPROVED** [review]
👍 Thanks @jonas-atmire !  This looks good to me now, and it fixes the problem.  I can now see mapped items within Collections from the UI.

**Comment by tdonohue** [comment]
@jonas-atmire : If you rebase this on `main`, the strange IT failures are now fixed (by #1491

**Comment by jonas-atmire** [comment]
@tdonohue Thanks for this quick fix.

We've subsequently updated the code related to this actual PR with your provided test-updates.
The tests now seem to be completing correctly again.

**Review by abollini — APPROVED** [review]

**Comment by abollini** [comment]
changes look good, thanks @jonas-atmire merging now

**Pull request merged**